pkg/envoy/lds/gateway.go

```diff
@@ -1,87 +1,80 @@
-			FilterChains: []*xds_listener.FilterChain{filterChain},
+	return &xds_listener.Listener{
+		Name:         multiclusterListenerName,
+		Address:      envoy.GetAddress(constants.WildcardIPAddr, constants.EnvoyMulticlusterListenerPort),
+		FilterChains: filterChains,
+		ListenerFilters: []*xds_listener.ListenerFilter{
+			{
+				Name: wellknown.TlsInspector,
+			},
 		},
-	}
+	}, nil
 }
 
-func getGatewayFilterChain(svcIdent identity.ServiceIdentity) (*xds_listener.FilterChain, error) {
-	tcpProxy := &xds_tcp_proxy.TcpProxy{
-		StatPrefix:       envoy.MulticlusterGatewayCluster,
-		ClusterSpecifier: &xds_tcp_proxy.TcpProxy_Cluster{Cluster: envoy.MulticlusterGatewayCluster},
-		AccessLog:        envoy.GetAccessLog(),
-	}
-	marshalledTCPProxy, err := ptypes.MarshalAny(tcpProxy)
-	if err != nil {
-		log.Error().Str(constants.LogFieldContext, constants.LogContextMulticluster).Err(err).Msg("[Multicluster] Error marshalling tcpProxy object for gateway filter chain")
-		return nil, err
-	}
+func getGatewayFilterChains(upstreamServices []service.MeshService) ([]*xds_listener.FilterChain, error) {
+	var filterChains []*xds_listener.FilterChain
+	for _, upstreamSvc := range upstreamServices {
+		tcpProxy := &xds_tcp_proxy.TcpProxy{
+			StatPrefix:       upstreamSvc.String(),
+			ClusterSpecifier: &xds_tcp_proxy.TcpProxy_Cluster{Cluster: upstreamSvc.String()},
+			AccessLog:        envoy.GetAccessLog(),
+		}
 
-	marshalledDownstreamTLSContext, err := ptypes.MarshalAny(envoy.GetDownstreamTLSContext(svcIdent, true /* mTLS */))
-	if err != nil {
-		log.Error().Str(constants.LogFieldContext, constants.LogContextMulticluster).Err(err).Msgf("[Multicluster] Error marshalling DownstreamTLSContext object for service identity %s", svcIdent)
-		return nil, err
-	}
+		marshalledTCPProxy, err := ptypes.MarshalAny(tcpProxy)
+		if err != nil {
+			log.Error().Err(err).Msgf("[Multicluster] Error marshalling tcpProxy object for gateway filter chain service %s", upstreamSvc.String())
+			continue
+		}
 
-	filterChain := &xds_listener.FilterChain{
-		Name: inboundMulticlusterGatewayFilterChainName,
-		FilterChainMatch: &xds_listener.FilterChainMatch{
-			ServerNames: []string{
-				"*.local",
-			},
-			TransportProtocol:    envoy.TransportProtocolTLS,
-			ApplicationProtocols: envoy.ALPNInMesh, // in-mesh proxies will advertise this, set in UpstreamTlsContext
-		},
-		Filters: []*xds_listener.Filter{
-			{
-				// This is of utmost importance to the Multicluster Gateway!
-				Name: "envoy.filters.network.sni_cluster",
-			},
-			{
-				Name: wellknown.TCPProxy,
-				ConfigType: &xds_listener.Filter_TypedConfig{
-					TypedConfig: marshalledTCPProxy,
+		filterChain := &xds_listener.FilterChain{
+			Name: fmt.Sprintf("%s-%s", multiclusterGatewayFilterChainName, upstreamSvc.Name),
+			FilterChainMatch: &xds_listener.FilterChainMatch{
+				// TODO : in the future .global and other cluster domains will be added here
```

Comment:
Is this true? If yes, would you mind changing the comment to `// TODO(#issue-num): ...`

Reply:
I am unsure at this point and we need to explore this a bit more, I have removed the comment for now